### PR TITLE
Refactor Minitest reporters to share code with future RSpec formatters

### PR DESCRIPTION
**Pull request by casperisfine**
This is the first of several refactors to make it easier to implement centralized reporting for RSpec.

The main goal here is that our custom minitest reporter stops talking to redis directly, but use a more high level interface that will be shared with RSpec. It has the added advantage of hiding wether we're using Redis, another backend, or a local in memory queue.

### CODE REVIEW

File: ruby/lib/ci/queue/build_record.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        record_stats(stats)
+      end
+
+      def fetch_stats(stat_names)
+        stat_names.zip(stats.values_at(*stat_names).map(&:to_f))
+      end
+
+      private
+
+      attr_reader :stats
```
Comment (DazWorrall) [inline ruby/lib/ci/queue/build_record.rb]: Nitpick: could this be merged with the `attr_reader` at the top of the class?

File: ruby/lib/ci/queue/redis/build_record.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+          @config = config
+        end
+
+        def progress
+          @queue.progress
+        end
+
+        def queue_exhausted?
+          @queue.exhausted?
+        end
```
Comment (DazWorrall) [inline ruby/lib/ci/queue/redis/build_record.rb]: There is some duplication in the methods above with `CI::Queue::BuildRecord`, could we inherit from it?

File: ruby/lib/ci/queue/build_record.rb (hunk 1 above)
Comment (casperisfine) [reply ruby/lib/ci/queue/build_record.rb]: No because I want ti to be private.

File: ruby/lib/ci/queue/redis/build_record.rb (hunk 2 above)
Comment (casperisfine) [reply ruby/lib/ci/queue/redis/build_record.rb]: Yeah we could probably share a base class or a module. I'll keep that in mind of a future refactor.

**Pull request merged**